### PR TITLE
Service expose/unexpose

### PR DESCRIPTION
**Pull request by josegonzalez**
This doesn't actually work because the iptables calls are wrong.

We should [link via an ambassador container](https://docs.docker.com/articles/ambassador_pattern_linking/).

/cc @Flink

### CODE REVIEW

**Comment by Flink** [comment]
I think it might need a little cleanup but it seems to work properly :)
Currently the `expose` command doesn’t take any argument I think so it’ll have to change.

**Comment by josegonzalez** [comment]
What happens if you run `redis:start` multiple times?

**Comment by Flink** [comment]
It doesn’t work as expected, there’s no check for `start` or `stop`. I focused on expose/unexpose ;)
Still, it needs some additional work, there is no autostart either.

**Comment by Flink** [comment]
Ok so now `start` and `stop` seem to work as expected. Autostart is working too. I also fixed a problem with `destroy`.
It still needs a big cleanup/refactor I think but there’s progress :)

**Comment by Flink** [comment]
Ok I think we’re good :smiley: I’ve squashed my commits and wrote a better one instead. If you want me to squash your commit too, just tell me :wink:

**Comment by progrium** [comment]
Why was ambassador pattern mentioned?

**Comment by josegonzalez** [comment]
@progrium in the context of exposing container ports on already running containers. You can't do that without doing iptables hacks, proxying on the host env, or the ambassador pattern.

**Comment by josegonzalez** [comment]
@Flink 

- Any thoughts on how we might be able to test this? Out of scope, but would be cool if you investigated that.
- Can you add a `service:port` command, so that people can see exposed ports?
- Is there a way we can expose a list of ports? That way services that have admin panels (rethinkdb) or multiple ports (mongodb) can have all their ports exposed to users.

For the last one, we can have the following in the `config` file of each service:

```shell
# comma-separated list of ports
PLUGIN_PORTS=6379,6380
```

We can then have a `PORTS` directory containing a `NUM` file. `NUM` would be the number of the port, and the contents of the file would be the publicly exposed mapping. Thoughts?

**Comment by Flink** [comment]
@josegonzalez 
* No I don’t know how to test, I think with bats? I’ll take a look on how to do it
* Sure thing! We could also display the exposed port via the `list` command
* Hmmm I don’t know, is this something really necessary? Does an admin panel is useful when we create a container per DB? And why not deploying an app that links to the wanted DB instead? (just asking)

**Comment by josegonzalez** [comment]
- Don't worry about testing now, that can be a separate pr.
- Awesome, have at it :)
- Yes, as there are some services that use multiple ports for non-admin (mongodb, elasticsearch). So if we can just support that pattern, would be *great*. One less container to manage.

**Comment by Flink** [comment]
Ok then, I’ll look how to do all this ;)

**Comment by Flink** [comment]
Ok I think it’s done. I tested several scenarios/use case but you should test it too.
To define exposed ports we only need to define it in `config`: `export PLUGIN_DATASTORE_PORTS=(6379)`. For another port just write it like this: `export PLUGIN_DATASTORE_PORTS=(6379 7000)`.

When calling the `expose` command without any args, the correct number of ports will be randomly generated (as expected). For custom ports, just provide them in the same order they are declared in `PLUGIN_DATASTORE_PORTS`. I added a check to be sure to provide the correct number of ports.

And the `list` command displays the exposed ports when there are ones :)

File: commands [hunk 1]
```diff
@@ -35,7 +35,8 @@ case "$1" in
     touch "$LINKS_FILE"
 
     dokku_log_info1 "Starting container"
-    ID=$(docker run --name "dokku.redis.$SERVICE" -v "$SERVICE_ROOT/data:/data" -v "$SERVICE_ROOT/config:/usr/local/etc/redis" -d "$PLUGIN_IMAGE:$PLUGIN_IMAGE_VERSION")
+    SERVICE_NAME=$(get_service_name "$SERVICE")
+    ID=$(docker run --name "$SERVICE_NAME" -v "$SERVICE_ROOT/data:/data" -v "$SERVICE_ROOT/config:/usr/local/etc/redis" -d --restart always "$PLUGIN_IMAGE:$PLUGIN_IMAGE_VERSION")
```
Comment (josegonzalez) [inline commands]: Why did you add `--restart always`?

File: commands [hunk 2]
```diff
@@ -73,10 +74,7 @@ case "$1" in
       dokku_log_verbose_quiet "Deleting container data"
       docker exec -it "$ID" chmod -R 777 /data
 
-      dokku_log_verbose_quiet "Stopping container"
-      docker stop "$ID" > /dev/null
-      docker kill "$ID" > /dev/null
-      sleep 1
+      service_stop "$SERVICE"
```
Comment (josegonzalez) [inline commands]: `service_stop` doesn't kill the container...

File: install [hunk 3]
```diff
@@ -8,3 +8,15 @@ fi
...
+  ubuntu)
+    echo "%dokku ALL=(ALL) NOPASSWD:/sbin/iptables -t nat -A DOCKER -p tcp *, /sbin/iptables -t nat -D DOCKER -p tcp *" > /etc/sudoers.d/dokku-redis
+    ;;
+
+  opensuse)
+    echo "%dokku ALL=(ALL) NOPASSWD:/sbin/iptables -t nat -A DOCKER -p tcp *, /sbin/iptables -t nat -D DOCKER -p tcp *" > /etc/sudoers.d/dokku-redis
+    ;;
+esac
+
+chmod 0440 /etc/sudoers.d/dokku-redis
```
Comment (josegonzalez) [inline install]: We don't need this anymore.

File: commands [hunk 4]
```diff
@@ -216,8 +202,8 @@ case "$1" in
     $PLUGIN_COMMAND_PREFIX:info <name>, Print the connection information
     $PLUGIN_COMMAND_PREFIX:list, List all $PLUGIN_SERVICE services
     $PLUGIN_COMMAND_PREFIX:clone <name> <new-name>, NOT IMPLEMENTED
-    $PLUGIN_COMMAND_PREFIX:expose <name> <port>, NOT IMPLEMENTED
-    $PLUGIN_COMMAND_PREFIX:unexpose <name> <port>, NOT IMPLEMENTED
+    $PLUGIN_COMMAND_PREFIX:expose <name> [port], Expose a $PLUGIN_SERVICE service on custom port if provided (random port otherwise)
+    $PLUGIN_COMMAND_PREFIX:unexpose <name>, Unexpose a previously exposed $PLUGIN_SERVICE service
```
Comment (josegonzalez) [inline commands]: Update the readme with docs for this?

Comment (josegonzalez) [inline commands]: Also needs `start`, `stop`, and `restart`.

Comment (josegonzalez) [inline commands]: I took care of this one

File: commands (hunk 1 above)
Comment (josegonzalez) [inline commands]: Hmm got it, it's for the start on reboot I guess.

**Comment by josegonzalez** [comment]
Okay so it looks like if you do the following:

```
dokku redis:stop lolipop
dokku cleanup # or deploy an app
```

The redis container disappears. I'll take care of this somehow upstream...

File: commands (hunk 2 above)
Comment (Flink) [inline commands]: You don't need to kill the container since `docker stop` will do it after a grace period (10 seconds by default). And you can't kill a stopped container, docker will output an error.

File: commands (hunk 1 above)
Comment (Flink) [inline commands]: Yes as explained here: https://github.com/dokku/dokku-postgres/issues/1#issuecomment-136971240 ;)

**Comment by Flink** [comment]
One way to fix this is to create the container on `start` if it doesn't exist

**Comment by josegonzalez** [comment]
If you're on IRC we can have a bit of a chat about this :)

**Comment by josegonzalez** [comment]
The following PR to dokku (https://github.com/progrium/dokku/pull/1444) should make the containers stick around through a `dokku cleanup`. One side-benefit is that users can also use the `dokku` label to create volumes that won't disappear :).

**Comment by josegonzalez** [comment]
progrium/dokku#1444 has landed, so this should work in latest master.

**Comment by josegonzalez** [comment]
Going to merge this awesomeness in now. Thanks so much @Flink for your work on this :+1:

**Comment by Flink** [comment]
You’re welcome @josegonzalez :) Sorry for not being on IRC, I was going to bed :grin: 
I’ll still have a PR to submit though!
